pkg/agent/plugin/workloadattestor/k8s/k8s.go

```diff
@@ -93,59 +102,115 @@ func (p *k8sPlugin) Attest(req *workloadattestor.AttestRequest) (*workloadattest
-	if err != nil {
-		return &resp, err
+	// Poll pod information and search for the pod with the container. If
+	// the pod is not found, and there are pods with containers that aren't
+	// fully initialized, delay for a little bit and try again.
+	for attempt := 1; ; attempt++ {
+		podInfo, err := p.getPodInfoFromInsecureKubeletPort()
+		if err != nil {
+			return &resp, err
+		}
+
+		notAllContainersReady := false
+		for _, item := range podInfo.Items {
+			switch lookUpContainerInPod(containerID, item.Status) {
+			case containerInPod:
+				resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("sa:%v", item.Spec.ServiceAccountName)})
+				resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("ns:%v", item.Metadata.Namespace)})
+				return &resp, nil
+			case containerMaybeInPod:
+				notAllContainersReady = true
+			case containerNotInPod:
+			}
+		}
+
+		// if the container was not located and there were no pods with
+		// uninitialized containers, then the search is over.
+		if !notAllContainersReady || attempt >= p.maxPollAttempts {
+			log.Printf("container id %q not found (attempt %d of %d)", containerID, attempt, p.maxPollAttempts)
+			return &resp, fmt.Errorf("no selectors found")
+		}
+
+		// wait a bit for containers to initialize before trying again.
+		log.Printf("container id %q not found (attempt %d of %d); trying again in %s", containerID, attempt, p.maxPollAttempts, p.pollRetryInterval)
+
+		// TODO: bail early via context cancelation
+		time.Sleep(p.pollRetryInterval)
 	}
-	defer httpResp.Body.Close()
-	respBytes, err := ioutil.ReadAll(httpResp.Body)
+}
 
-	var podInfo *podList
-	err = json.Unmarshal(respBytes, &podInfo)
+func (p *k8sPlugin) getPodInfoFromInsecureKubeletPort() (out *podList, err error) {
+	httpResp, err := p.httpClient.Get(fmt.Sprintf("http://localhost:%d/pods", p.kubeletReadOnlyPort))
 	if err != nil {
-		return &resp, err
+		return nil, err
 	}
+	defer httpResp.Body.Close()
 
-	for _, item := range podInfo.Items {
-		match, err := statusMatches(containerID, item.Status)
-		if err != nil {
-			return &resp, err
-		}
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", httpResp.StatusCode)
+	}
 
-		if match {
-			resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("sa:%v", item.Spec.ServiceAccountName)})
-			resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("ns:%v", item.Metadata.Namespace)})
-			return &resp, nil
-		}
+	out = new(podList)
+	if err := json.NewDecoder(httpResp.Body).Decode(out); err != nil {
+		return nil, err
 	}
 
-	return &resp, fmt.Errorf("no selectors found")
+	return out, nil
 }
 
-func statusMatches(containerID string, status podStatus) (bool, error) {
+type containerLookup int
+
+const (
+	containerInPod = iota
+	containerNotInPod
+	containerMaybeInPod
```

Comment:
Nice

Is it idiomatic to keep these defined close to the function which uses them, or at the top of the file? Feels like the latter - that's where I went looking when I came across them in code above.

Reply:
For me it has always come down to scope. Since these were just used in one function it made sense to me to locate them closely. I don't have a strong opinion on this and certainly top of the file is a natural fit either way. Happy to move them.